### DISCUSSION TITLE
JavaScript interop wit Avalonia WASM

### DISCUSSION THREAD

**Discussion by pholpar** [AvaloniaUI/Avalonia#8511] [Q&A] [answered]
Sorry if I overlooked an existing feature, but I have not found any option to call JavaScript functions from C# or to invoke C# methods from JavaScript when working with Avalonia in the Browser, similar to the way, as it works with Blazor:
https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability
https://www.imaginet.com/2021/guide-blazor-javascript-interop

Currently, I don't see how to perform such trivial steps, like to get the URL of the web page the Avalonia WASM component hosted on, that may be relevant if we would like to interact with an API that is served by the same webserver, or to download and display an image dynamically from the same server, without knowing the URL in advance, or using some kind of configuration. Not to mention some more advanced scenarios.

If this feature is available, please post a link to the page, where it is described.

// What I found a few days ago is the class JSModuleInterop (https://github.com/AvaloniaUI/Avalonia/blob/master/src/Web/Avalonia.Web.Blazor/Interop/JSModuleInterop.cs), but as it is defined as internal, we can probably not utilize its power from our code.

Thank you in advance!

**Comment by jp2masa** [comment, marked as answer]
#8017

Avalonia is hosted on a Blazor component, so you can use most of the Blazor APIs.

**Reply by pholpar** [reply]
Thanks for the quick answer! It was clear, that Avalonia in the background utilizes Blazor, just I didn't see, how to access its features. The referenced issue seems to provide the information I missed.

Sorry to bother you with a question already answered, but I used the search terms (JS, JavaScript, interop) that didn't give me the result I expected, and not the keyword 'IJSRuntime'.

**Comment by maxkatz6** [comment]
With 11.0 we will have Blazor-less integration. It is recommended to use .NET 7 standard build in interop via JSImport and JSExport attributes.

**Comment by johnnyprogrammer** [comment]
I am trying to port avalonia cross platform desktop app to web using WASM and Blazor and run into problem that I need javascript to call API in browser sandbox.

Q1: Is there a way to calla javascript interop for API   directly from MainView : UserControl without writing # code on  Razor page via either @inject IJSRuntime JS or JSImport/JSExport?

If Q1 is not possible, then I would guess that I have to some how inject html button for calling javascript API inside Index.razor along side with AvaloniaView?  Any hint or guide for this is very much appreciated.

No matter what, I am thanksful for both Avalonia and this thread for the direction. I am  a web, native mac/win desktop,  android/iOS developer and Avalonia seems to be a perfect match for me to unify these. I took risk learning Avalonia  for crossplatform desktop app last year and the performance beyond my expectation ( I obsessed over getting stuck and couldn't google out a way  and had to go with another framework).

**Reply by maxkatz6** [reply]
> Q1: Is there a way to calla javascript interop for API directly from MainView : UserControl without writing # code on Razor page via either @Inject IJSRuntime JS or JSImport/JSExport?

`AvaloniaLocator.Current.GetService<IJSInProcessRuntime>`

But it's not recommended to use Blazor bidnings anymore, as Microsoft finally released .NET WASM Interop in .NET 7. It is also faster interop comparing to Blazor, not to mention there is less abstraction layers.
You can use `Avalonia.Web` package from nuget (preview, 11.0-preview3) and simple example including interop can be found here
https://github.com/AvaloniaUI/Avalonia/tree/master/samples/ControlCatalog.Web

**Reply by maxkatz6** [reply]
More details about new interop from MS
https://devblogs.microsoft.com/dotnet/use-net-7-from-any-javascript-app-in-net-7/
